torch/lib/THD/base/data_channels/Store.cpp

```diff
@@ -0,0 +1,232 @@
+  }
+  _socket = connect(_master_addr, _master_port);
+  send_bytes<rank_type>(_socket, &_rank, 1);
+}
+
+Store::~Store() {
+  // the 0 process has to stop the daemon
+  query_type fin = query_type::finish;
+  if (_rank == 0) {
+    send_bytes<query_type>(_socket, &fin, 1);
+    _store_thread.join();
+  }
+}
+
+void Store::set(const std::string& key, const std::vector<char>& data) {
+  send_query(_socket, query_type::set);
+  send_string(_socket, key, true);
+  send_vector<char>(_socket, data);
+}
+
+std::vector<char> Store::get(const std::string& key) {
+  wait({key});
+  send_query(_socket, query_type::get);
+  send_string(_socket, key);
+  return recv_vector<char>(_socket);
+}
+
+void Store::wait(const std::vector<std::string>& keys) {
+  send_query(query_type::wait);
+  size_type nkeys = keys.size();
+  send_bytes<size_type>(_socket, &nkeys, 1, (nkeys > 0));
+  for (int i = 0; i < nkeys; i++) {
+    send_string(_socket, keys[i], (i != (nkeys - 1)));
+  }
+  // after sending the query, wait for a 'stop_waiting' reponse
+  query_type qr;
+  recv_bytes<query_type>(_socket, &query, 1);
+  if (query != query_type::stop_waiting)
+    throw std::runtime_error("stop_waiting response expected");
+}
+
+static void store_thread_daemon() {
+  port_type port;
+  rank_type world_size;
+  int socket;
+  std::tie(port, world_size) = load_master_env();
+  std::vector<int> sockets(world_size);
+  std::tie(socket, std::ignore) = listen(port);
+  // accept WORLD_SIZE connections
+  for (int i = 0; i < world_size; i++) {
+    int p_socket;
+    rank_type p_rank;
+    std::tie(p_socket, std::ignore) = accept(socket);
+    recv_bytes<rank_type>(p_socket, &p_rank, 1);
+    sockets[p_rank] = p_socket;
+  }
+
+  // listen for requests
+  struct pollfd fds[world_size];
+  for (int i = 0; i < world_size; i++) {
+    fds[i].fd = sockets[i];
+    fds[i].events = POLLIN;   // we only read
+  }
+
+  // receive the queries
+  bool finished = false;
+  while (!finished) {
+    for (int i = 0; i < world_size; i++) fds[i].revents = 0;
+    SYSCHECK(poll(fds, world_size, -1));
+    for (int i = 0; i < world_size; i++) {
+      if (fds[i].revents == 0)
+        continue;
+      if (fds[i].revents ^ POLLIN)
+        throw std::system_error(ECONNABORTED, std::system_category());
+      finished = query(fds[i].fd);
+      if (finished)
+        break;
+    }
+  }
+}
+
+/* 
+ * query communicates with the worker. The format
+ * of the query is as follows:
+ * type of query | size of arg1 | arg1 | size of arg2 | arg2 | ...
+ * or, in the case of wait
+ * type of query | number of args | size of arg1 | arg1 | ...
+ */
+static bool query(int socket) {
+  static store_type store_;
+  static std::unordered_map<int, std::vector<std::string>> waiting;
+  query_type qt;
+  recv_bytes<query_type>(socket, &qt, 1);
+  if (qt == query_type::set) {
+    std::string key = recv_string(socket);
+    std::vector<char> data = recv_vector<char>(socket);
+    store_[key] = std::move(data);
```

Comment:
You can just do `store_[key] = recv_vector<char>(socket);` and achieve same result. But not sure if this is as clear as your code.